modules/packages/LinearAlgebra.chpl

```diff
@@ -841,4 +843,467 @@ proc trace(A: [?D] ?eltType) {
+  D += (2,2);
+
+  // Set all nonzero indices to the value of 1
+  A = 1;
+
+  // A is now a 3x3 sparse identity matrix
+  writeln(A);
+
+
+.. note::
+  This is an early prototype package submodule. As a result, interfaces may
+  change over the next release.
+
+*/
+module Sparse {
+
+  use LayoutCS;
+
+  /* Return an empty CSR domain over parent domain:
+     ``{0..#rows, 0..#rows}``
+   */
+  proc CSRDomain(rows) where isIntegral(rows) {
+    if rows <= 0 then halt("Matrix dimensions must be > 0");
+    return CSRDomain(0..#rows, 0..#rows);
+  }
+
+
+  /* Return an empty CSR domain  over parent domain: ``{0..#rows, 0..#cols}``*/
+  proc CSRDomain(rows, cols) where isIntegral(rows) && isIntegral(cols) {
+    if rows <= 0 || cols <= 0 then halt("Matrix dimensions must be > 0");
+    return CSRDomain(0..#rows, 0..#cols);
+  }
+
+
+  /* Return an empty CSR domain over parent domain: ``{space, space}`` */
+  proc CSRDomain(space: range) {
+    return CSRDomain({space, space});
+  }
+
+
+  /* Return an empty CSR domain over parent domain: ``{rowSpace, colSpace}`` */
+  proc CSRDomain(rowSpace: range, colSpace: range) {
+    return CSRDomain({rowSpace, colSpace});
+  }
+
+  /* Return a CSR domain based on domain: ``Dom``
+
+    If ``Dom`` is dense, it will be interpreted as the parent domain, and the
+    domain returned will be empty.
+
+    If ``Dom`` is sparse (CSR), the domain returned will contain the same
+    nonzeros as ``Dom``
+  */
+  proc CSRDomain(Dom: domain) where Dom.rank == 2 && isCSDom(Dom) {
+    var csrDom: sparse subdomain(Dom._value.parentDom) dmapped CS();
+    csrDom += Dom;
+    return csrDom;
+  }
+
+  pragma "no doc"
+  /* Return a CSR domain based on domain: ``Dom`` - Dense case */
+  proc CSRDomain(Dom: domain(2)) where Dom.rank == 2 {
+    var csrDom: sparse subdomain(Dom) dmapped CS();
+    return csrDom;
+  }
+
+
+  /* Return a CSR matrix over domain: ``Dom``
+
+    If ``Dom`` is dense, it will be interpreted as the parent domain, and the
+    matrix returned will be empty.
+
+    If ``Dom`` is sparse (CSR), the matrix returned will contain the same
+    nonzeros as ``Dom``
+  */
+  proc CSRMatrix(Dom: domain, type eltType=real) where Dom.rank == 2 && isCSDom(Dom) {
+    var M: [Dom] eltType;
+    return M;
+  }
+
+  pragma "no doc"
+  /* Return a CSR matrix over domain: ``Dom`` - Dense case */
+  proc CSRMatrix(Dom: domain, type eltType=real) where Dom.rank == 2 && isDefaultDom(Dom) {
+    var csrDom = CSRDomain(Dom);
+    var M: [csrDom] eltType;
+    return M;
+  }
+
+
+  /* Return a CSR matrix with domain and values of ``A``
+
+    ``A`` can be a dense or sparse matrix.
+   */
+  proc CSRMatrix(A: [?Dom] ?Atype, type eltType=Atype) where Dom.rank == 2 && isCSArr(A) {
+    var M: [Dom] eltType = A: eltType;
+    return M;
+  }
+
+  pragma "no doc"
+  /* Return a CSR matrix with domain and values of ``A`` - Dense case */
+  proc CSRMatrix(A: [?Dom] ?Atype, type eltType=Atype) where Dom.rank == 2 && isDefaultArr(A) {
```

Comment:
What's the value in having a sparse array that is all nonzeroes?

Reply:
When given a dense array, only the indices holding nonzero elements are added to the sparse array. I'll update the documentation to make this clearer.

Reply:
Ah, gotcha.